todoman/model.py

```diff
@@ -231,74 +274,493 @@ def _normalize_datetime(self, x):
+        ).fetchone()
+
+        return result['id']
+
+    def add_todo(self, todo, file_id):
+        """
+        Adds a todo into the cache.
+
+        :param icalendar.Todo todo: The icalendar component object on which
+        """
+
+        self.cur.execute('''
+            INSERT INTO todos (
+                file_id,
+                uid,
+                summary,
+                due,
+                priority,
+                created_at,
+                completed_at,
+                percent_complete,
+                dtstamp,
+                status,
+                description,
+                location,
+                categories
+            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+            ''', (
+                file_id,
+                todo.uid,
+                todo.summary,
+                todo.due,
+                todo.priority,
+                todo.created_at,
+                todo.completed_at,
+                todo.percent_complete,
+                todo.dtstamp,
+                todo.status,
+                todo.description,
+                todo.location,
+                todo.raw_categories,
+            )
+        )
+
+        self.conn.commit()
+
+    def todos(self, all=False, lists=[], urgent=False, location='',
+              category='', grep='', sort=[], reverse=False):
+        todos = []
+        list_map = self._lists_map()
+
+        extra_where = []
+        params = []
+
+        if not all:
+            extra_where.append('AND completed_at is NULL')
+        if lists:
+            q = ', '.join(['?'] * len(lists))
+
+            query = 'SELECT id FROM lists WHERE name IN ({});'.format(q)
+            list_ids = [row['id'] for row in self.conn.execute(q, lists)]
+            extra_where.append('AND files.list_id IN ({})'.format(q))
+            params.extend(list_ids)
+        if urgent:
+            extra_where.append('AND priority = 9')
+        if location:
+            extra_where.append('AND location LIKE ?')
+            params.append('%{}%'.format(location))
+        if category:
+            extra_where.append('AND categories LIKE ?')
+            params.append('%{}%'.format(category))
+        if grep:
+            # # requires sqlite with pcre, which won't be availabel everywhere:
+            # extra_where.append('AND summary REGEXP ?')
+            # params.append(grep)
+            extra_where.append('AND summary LIKE ?')
+            params.append('%{}%'.format(grep))
+
+        if sort:
+            order = []
+            for s in sort:
+                if s.startswith('-'):
+                    order.append(' {} ASC'.format(s[1:]))
+                else:
+                    order.append(' {} DESC'.format(s))
+            order = ','.join(order)
+        else:
+            order = '''
+                completed_at DESC,
+                priority ASC,
+                due IS NOT NULL, due DESC,
+                created_at ASC
+            '''
+
+        if not reverse:
```

Comment:
If `not reverse` is the more common codepath it may make sense to generate the "not-reverse" query by default, and apply a text-replacement only if `reverse=True`.